### PR TITLE
New module: Gitlab service

### PR DESCRIPTION
**Pull request by drzraf**
##### SUMMARY
New module: **Gitlab integration services**

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
gitlab_service

##### ANSIBLE VERSION
```
ansible 2.5.2
```

##### ADDITIONAL INFORMATION
This module provides a way to setup Gitlab integration services using Ansible.
Contrary to the first 3 gitlab modules (project, user and group), it relies upon the python-gitlab module instead of pyapi-gitlab (see #29310) 

This module theorically provides support to all services defined by Gitlab.com API on 2018/05/12 (and their corresponding suboptions) as long as they are supported by the python-gitlab wrapper (1.3.0 as of writing)

Known bugs:
- The _active_ state of a service [can not be set using the API](https://gitlab.com/gitlab-org/gitlab-ce/issues/41113). Use `state=absent` to completely remove the service configuration instead of deactivating it.
- Most Gitlab integration services rely upon _secret_ parameter (password, token, api_key) which are write-only (ie: not exposed on GET request). That currently [keeps module from being idempotent when such a suboption is used](https://gitlab.com/gitlab-org/gitlab-ce/issues/46313).

##### EXTENDED DEFINITION OF SERVICE
- Their validation is done by doing a redundant initialization of the `AnsibleModule()` and modifying the argument spec meanwhile.

- The extended definition of services and their suboptions (**name**, **required** or not, **type**) is generated from [Gitlab upstream file](https://gitlab.com/gitlab-org/gitlab-ee/blob/f850b1bdb74de78875a49126e0401cc975fda32a/lib/api/services.rb) (adding a couple of Ruby loops at the end of the Ruby class definition):

```ruby
    ###
    services.each do |service_slug, params|
      new_settings = {}

      params.each do |entry|
        e = {}

        if (entry[:required] === true)
          e[:required] = true
        end
        if (entry[:type] === "String")
        elsif (entry[:type] === "Integer")
          e[:type] = "int"
        else
          e[:type] = entry[:type]
        end
        # e[:desc] = entry[:desc]

        new_settings[entry[:name]] = e
      end
      services[service_slug] = new_settings
    end
    puts JSON.generate(services)

### CODE REVIEW

**Comment by dj-wasabi** [comment]
Looks good to me.

shipit

**Comment by Shaps** [comment]
@drzraf Thanks for submitting this new module, can you please add some integration tests for it? 
You can use #51490 as a reference for what integration tests look like

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: gitlab_service
+short_description: Creates, updates or deletes Gitlab services
```
Comment (adrianmoisey) [inline lib/ansible/modules/source_control/gitlab_service.py]: GitLab prefer to be called `GitLab` (with a capital `L`)

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+version_added: "2.6"
+description:
+  - Setup or delete Gitlab integration services.
+author: "Raphaël Droz (@drzraf)"
+requirements:
+  - python-gitlab
+options:
+  server_url:
+    description:
+      - Url of Gitlab server, with protocol (http or https).
```
Comment (adrianmoisey) [inline lib/ansible/modules/source_control/gitlab_service.py]: `URL`

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+    required: true
+  validate_certs:
+    description:
+      - When using https if SSL certificate needs to be verified.
```
Comment (adrianmoisey) [inline lib/ansible/modules/source_control/gitlab_service.py]: I'd reword this as `Verify SSL certificate when HTTPS is used`

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
...,"merge_request_channel":{},"note_channel":{},"tag_push_channel":{},"pipeline_channel":{},"wiki_page_channel":{},"push_events":{"type":"bool"},"issues_events":{"type":"bool"},"confidential_issues_events":{"type":"bool"},"merge_requests_events":{"type":"bool"},"note_events":{"type":"bool"},"tag_push_events":{"type":"bool"},"pipeline_events":{"type":"bool"},"wiki_page_events":{"type":"bool"}},"microsoft-teams":{"webhook":{"required":true}},"mattermost":{"webhook":{"required":true},"username":{},"channel":{},"notify_only_broken_pipelines":{"type":"bool"},"notify_only_default_branch":{"type":"bool"},"push_channel":{},"issue_channel":{},"confidential_issue_channel":{},"merge_request_channel":{},"note_channel":{},"tag_push_channel":{},"pipeline_channel":{},"wiki_page_channel":{},"push_events":{"type":"bool"},"issues_events":{"type":"bool"},"confidential_issues_events":{"type":"bool"},"merge_requests_events":{"type":"bool"},"note_events":{"type":"bool"},"tag_push_events":{"type":"bool"},"pipeline_events":{"type":"bool"},"wiki_page_events":{"type":"bool"}},"teamcity":{"teamcity_url":{"required":true},"build_type":{"required":true},"username":{"required":true},"password":{"required":true}}}
```
Comment (Lunik) [inline lib/ansible/modules/source_control/gitlab_service.py]: This definition hasn't been updated since a long time.
Just made a quick check, there are new services :  `discord`, `hangouts-chat`.
We need a more recent version of this dump

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+    active = True
+    # active = module.params['active']
+    params = module.params['params']
+    events = module.params['events']
+
+    try:
+        git = Gitlab(url=api_url, ssl_verify=validate_certs, email=api_username, password=api_password, private_token=api_token, api_version=4)
+        # git.enable_debug() ?
+        git.auth()
+        project = git.projects.get(project)
```
Comment (Lunik) [inline lib/ansible/modules/source_control/gitlab_service.py]: Split GitLab authentification en project gathering.
Add a custom check on `git.projects.get(project)` with a custom message

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+                module.exit_json(changed=True, service=remote_service.attributes, result='Successfully deleted service %s' % service)
+
+    else:
+        if remote_service.created_at:
+            # update
+            try:
+                diff = {'before': str(remote_service.attributes)}
+                h = ServicesHelper.update(project, remote_service, active, params, events)
+                # if diff: fetch again
+                # diff['after'] = str(remote_service.attributes)
```
Comment (Lunik) [inline lib/ansible/modules/source_control/gitlab_service.py]: Why is this code commented ?
Don't you want to return diff after ?

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+                                     result='Successfully updated service %s' % service)
+                else:
+                    module.exit_json(changed=False)
+
+        else:
+            # create (could only happen once for this service during the life of the project)
+            # merge new attributes in the object retrieved from the server
+            local_service = ServicesHelper.asApiObject(active, params, events)
+            for k, v in local_service.items():
+                setattr(remote_service, k, v)
```
Comment (Lunik) [inline lib/ansible/modules/source_control/gitlab_service.py]: Create a new method on the `GitLabServices` class to do the creation + saving job

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+    except gitlab.GitlabGetError as e:
+        module.fail_json(msg='No such a service %s' % service, exception=to_native(e))
+
+    ServicesHelper = GitLabServices(module, git)
+    if state == 'absent':
+        if not remote_service or not remote_service.created_at:
+            module.exit_json(changed=False, service={}, msg='Service not found', details='Service %s not found' % service)
+        else:
+            if module.check_mode:
+                module.exit_json(changed=True, service=remote_service.attributes)
```
Comment (Lunik) [inline lib/ansible/modules/source_control/gitlab_service.py]: The return msg should be indentical in normal and check mode

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+            try:
+                remote_service.delete()
+            except (gitlab.GitlabHttpError, gitlab.GitlabDeleteError) as e:
+                module.fail_json(msg='Failed to remove service %s' % service, exception=to_native(e))
+            else:
+                module.exit_json(changed=True, service=remote_service.attributes, result='Successfully deleted service %s' % service)
```
Comment (Lunik) [inline lib/ansible/modules/source_control/gitlab_service.py]: Use `msg` instead of `result` attribute
```suggestion
                module.exit_json(changed=True, service=remote_service.attributes, msg='Successfully deleted service %s' % service)
```

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+                diff = {'before': str(remote_service.attributes)}
+                h = ServicesHelper.update(project, remote_service, active, params, events)
+                # if diff: fetch again
+                # diff['after'] = str(remote_service.attributes)
+            except gitlab.GitlabUpdateError as e:
+                module.fail_json(changed=False, msg='Could not update service %s' % service, exception=to_native(e))
+            else:
+                if h:
+                    module.exit_json(changed=True, service=remote_service.attributes, diff=diff, state='changed',
+                                     result='Successfully updated service %s' % service)
```
Comment (Lunik) [inline lib/ansible/modules/source_control/gitlab_service.py]: Use `msg` instead of `result` attribute
```suggestion
                                     msg='Successfully updated service %s' % service)
```

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+            if module.check_mode:
+                module.exit_json(changed=True)
+            try:
+                remote_service.save()
+                diff = {'before': {}, 'after': str(remote_service.attributes)}
+            except (gitlab.GitlabCreateError, gitlab.GitlabUpdateError) as e:
+                module.fail_json(changed=False, msg='Could not create service %s' % service, exception=to_native(e))
+            else:
+                module.exit_json(changed=True, service=remote_service.attributes, diff=diff, state='created',
+                                 result='Successfully created service %s' % service)
```
Comment (Lunik) [inline lib/ansible/modules/source_control/gitlab_service.py]: Use `msg` instead of `result` attribute
```suggestion
                                 msg='Successfully created service %s' % service)
```

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+                # if diff: fetch again
+                # diff['after'] = str(remote_service.attributes)
+            except gitlab.GitlabUpdateError as e:
+                module.fail_json(changed=False, msg='Could not update service %s' % service, exception=to_native(e))
+            else:
+                if h:
+                    module.exit_json(changed=True, service=remote_service.attributes, diff=diff, state='changed',
+                                     result='Successfully updated service %s' % service)
+                else:
+                    module.exit_json(changed=False)
```
Comment (Lunik) [inline lib/ansible/modules/source_control/gitlab_service.py:440]: The return msg should be indentical in normal and check mode

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+      server: https://packagist.org
+  delegate_to: localhost
+
+- Idempotency is only partially provided since GitLab does
+  not expose secret parameters like tokens or passwords.
+  See U(https://gitlab.com/gitlab-org/gitlab-ce/issues/46313)
+'''
+
+RETURN = '''
+---
```
Comment (Lunik) [inline lib/ansible/modules/source_control/gitlab_service.py:132]: The `diff` return is not present in the documentation

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+  project:
+    description:
+      - The ID or path of the project (urlencoded or not)
+    required: true
+    default: null
+  active:
+    description:
+      - Not yet supported (U(https://gitlab.com/gitlab-org/gitlab-ce/issues/41113))
+    type: bool
+    default: 'yes'
```
Comment (Lunik) [inline lib/ansible/modules/source_control/gitlab_service.py]: Quote are not required here

```suggestion
    default: yes
```

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+        if diff:
+            # remote.active = active
+            for k, v in self.expandEvents(events).items():
+                setattr(remote, k, v)
+            # when saving, params does directly in the request
+            # (remember they get retrieved under the "property" key of the remote dict())
+            for k, v in params.items():
+                setattr(remote, k, v)
+            if self._module.check_mode:
+                self._module.exit_json(changed=True)
```
Comment (Lunik) [inline lib/ansible/modules/source_control/gitlab_service.py]: The return msg should be indentical in normal and check mode

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 16]
```diff
@@ -0,0 +1,377 @@
...
+            remote.save()
+            return remote
+        else:
+            return False
+
+    def asApiObject(self, active, params, events):
```
Comment (Lunik) [inline lib/ansible/modules/source_control/gitlab_service.py:293]: Why not using directly `project.services.create` method ?

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 17]
```diff
@@ -0,0 +1,377 @@
...
+        return ('password' in attr and attr['password']) or ('token' in attr and attr['token']) or ('api_key' in attr and attr['api_key'])
+
+    def equals(self, prev_attr, active, params, events):
+        # python >= 2.7: filtered_params = {k: v for k, v in params.items() if k not in self.credentialParams}
+        filtered_params = {}
+        for k, v in params.items():
+            if k not in self.credentialParams:
+                setattr(filtered_params, k, v)
+        # ToDo: this is not (yet?) supported by GitLab
+        # if prev_attr['active'] != active:
```
Comment (Lunik) [inline lib/ansible/modules/source_control/gitlab_service.py]: Is it now ?

**Review by Lunik — COMMENTED** [review]
There are some changes to make

File: lib/ansible/modules/source_control/gitlab_service.py (hunk 17 above)
Comment (drzraf) [reply lib/ansible/modules/source_control/gitlab_service.py]: still not, see https://gitlab.com/gitlab-org/gitlab-ce/issues/41113

File: lib/ansible/modules/source_control/gitlab_service.py (hunk 16 above)
Comment (drzraf) [reply lib/ansible/modules/source_control/gitlab_service.py:293]: As far as I remember I had some trouble using `create` without having the tweak the `dict()/data-structure` in use.
And since we always do `remote_service = project.services.get(service)`, we always have a `remote service` object suitable for `project.services.save`.

**Comment by drzraf** [comment]
Can't we just ignore remaining warnings because:
- Lines 193 is autogenerated. Line length won't hurt.
- python2.6 is not expected to be supported because it's not by python-gitlab which is required by this module and the 3 constructs which are >= python2.7 makes code cleaner.

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 18]
```diff
@@ -0,0 +1,400 @@
...
...ents": {"type": "bool"}}, "microsoft-teams": {"webhook": {"required": 1}}, "mattermost": {"webhook": {"required": 1}, "username": {}, "channel": {}, "notify_only_broken_pipelines": {"type": "bool"}, "notify_only_default_branch": {"type": "bool"}, "push_channel": {}, "issue_channel": {}, "confidential_issue_channel": {}, "merge_request_channel": {}, "note_channel": {}, "tag_push_channel": {}, "pipeline_channel": {}, "wiki_page_channel": {}, "push_events": {"type": "bool"}, "issues_events": {"type": "bool"}, "confidential_issues_events": {"type": "bool"}, "merge_requests_events": {"type": "bool"}, "note_events": {"type": "bool"}, "tag_push_events": {"type": "bool"}, "pipeline_events": {"type": "bool"}, "wiki_page_events": {"type": "bool"}}, "teamcity": {"teamcity_url": {"required": 1}, "build_type": {"required": 1}, "username": {"required": 1}, "password": {"required": 1}}, "github": {"token": {"required": 1}, "repository_url": {"required": 1}}, "jenkins": {"jenkins_url": {"required": 1}, "project_name": {"required": 1}, "username": {}, "password": {}}, "jenkins-deprecated": {"project_url": {"required": 1}, "pass_unstable": {"type": "bool"}, "multiproject_enabled": {"type": "bool"}}}
```
Comment (sivel) [inline lib/ansible/modules/source_control/gitlab_service.py]: This line should be "prettified" by adding indentation and line breaks, so that it properly fits within our line limits.

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 19]
```diff
@@ -0,0 +1,400 @@
...
+        return obj
+
+    def expandEvents(self, events):
+        try:
+            # See https://gitlab.com/gitlab-org/gitlab-ce/issues/58321 for why it's useful to
+            # discard unsupported events before comparing
+            supported_events = RAW_SERVICES_DEFINITIONS[self.name]['_events']
+        except KeyError:
+            supported_events = None
+        return {value + '_events': bool(value in events) for value in self.HOOK_EVENTS if supported_events is None or value in supported_events}
```
Comment (sivel) [inline lib/ansible/modules/source_control/gitlab_service.py:306]: Instead of a dict comprehension, which our CI has flagged as invalid, please use a `dict()` constructor with a generator expression:

```
dict((k, v) for k, v in thing.items())
```

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 20]
```diff
@@ -0,0 +1,400 @@
...
+
+    def eventsEqual(self, obj, events):
+        return all((k in obj and obj[k] == v) for k, v in events.items())
+
+    def hasCredentials(self, attr):
+        return ('password' in attr and attr['password']) or ('token' in attr and attr['token']) or ('api_key' in attr and attr['api_key'])
+
+    def equals(self, prev_attr, active, params, events):
+        filtered_params = {k: v for k, v in params.items() if k not in self.credentialParams}
```
Comment (sivel) [inline lib/ansible/modules/source_control/gitlab_service.py:315]: No dict comprehension here.

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 21]
```diff
@@ -0,0 +1,400 @@
...
+        ],
+        supports_check_mode=True
+    )
+
+    stub_init = AnsibleModule(**base_specs)
+    service = stub_init.params['service']
+    state = stub_init.params['state']
+    if state == 'present':
+        # since we know the service (which has been validated), recreate a module_specs to validate suboptions
+        sub_arg_specs = {i: definitions[service][i] for i in definitions[service] if i != '_events'}
```
Comment (sivel) [inline lib/ansible/modules/source_control/gitlab_service.py:373]: Another dict comprehension.

File: lib/ansible/modules/source_control/gitlab_service.py [hunk 22]
```diff
@@ -0,0 +1,400 @@
...
+        argument_spec=dict(
+            api_url=dict(required=True),
+            validate_certs=dict(required=False, default=True, type='bool'),
+            api_username=dict(required=False, no_log=True),
+            api_password=dict(required=False, no_log=True),
+            api_token=dict(required=False, no_log=True),
+            project=dict(required=True),
+            service=dict(required=False, type='str', choices=list(definitions.keys())),
+            active=dict(required=False, default=True, type='bool', choices=[True]),
+            params=dict(required=False, type='dict'),
```
Comment (sivel) [inline lib/ansible/modules/source_control/gitlab_service.py:348]: This argument violates our requirements for module arguments.  If you cannot enumerate all options this param takes, it violates our guidelines.

Please see item #5 at https://docs.ansible.com/ansible/latest/dev_guide/developing_modules.html

> Should you write multiple modules instead of one module?
> The functionality you want may be too large for a single module. If you want to connect Ansible to a new cloud provider, database, or network platform, you may need to develop a related group of modules.
> * Modules should have a concise and well defined functionality. Basically, follow the UNIX philosophy of doing one thing well.
> * Modules should not require that a user know all the underlying options of an API/tool to be used. For instance, if the legal values for a required module parameter cannot be documented, that’s a sign that the module would be rejected.
> * Modules should typically encompass much of the logic for interacting with a resource. A lightweight wrapper around an API that does not contain much logic would likely cause users to offload too much logic into a playbook, and for this reason the module would be rejected. Instead try creating multiple modules for interacting with smaller individual pieces of the API.

As such, this single module should become many modules that interact with individual components of the Gitlab API.

**Review by sivel — CHANGES_REQUESTED** [review]
This module violates our guidelines for developing a module, and should instead be submitted as many modules that interact with individual components of the Gitlab API.

If you cannot concisely document all options that an argument will accept, it cannot be accepted into this repository.

File: lib/ansible/modules/source_control/gitlab_service.py (hunk 19 above)
Comment (drzraf) [reply lib/ansible/modules/source_control/gitlab_service.py:306]: CI has flagged the Python 2.6 incompatible. As discussed before, this module depends upon Python 2.7+ for the reason that the underlying python-gitlab library is 2.7+
As such there does not seem that the >= 2.6 requirement for the PR itself applies here since the python-2.6 CI  green-light would just be misleading anyway.

**Comment by drzraf** [comment]
Let me recall : Among the functionalities provided by GitLab, a part is about _provisioning a project/group/..._ GitLab **Services are one tiny part** of this setup which is documented here : https://docs.gitlab.com/ee/user/project/integrations/project_services.html

The fact that a lot of different `brands` appear in the code of this module must not mislead anyone : Even if _Bugzilla_ is quite different from _Packagist_ at first glance, GitLab API (and this module)  provision one specific aspect **of GitLab** : **service integration**.

Creating 30 modules for one sole aspect would be a hurdle for everyone! (which what the guidelines want to avoid!)
Anyone who setup its GitLab services will *appreciate* much more having one versatile, comprehensive and nicely validated module to set this up instead of remembering and choosing among dozen of different ones. (not even speaking extra feature like loops or maintainer's difficulties track and semi-automatically generate individual validation that would imply that 30 modules will be less up-to-date than one).

I also have to add that I may have very well put `params` as an arbitrary and **unvalidated** object field, as **many other module already do** (under name like `opaque blob` or `user data`). Contrary to this deliberately went further and introduced validation at a deeper level than strictly needed in order to help users. As such I find very unfortunate that the effort spent to improve module quality beside the minimum-viable-module could be used to argue against its introduction. That's sounds as if [mount module](https://docs.ansible.com/ansible/latest/modules/mount_module.html#mount-module) authors would have written an extra-logic of validation of `opts` according to `fstype` and that reviewers would argue, against a merge, and advising to  split it among `mount_dos`, `mount_ext4`, ...

**Comment by Shaps** [comment]
Since this seems to be a bit controversial, I'd probably add it to the next Core meeting. @drzraf would you be able to make it ?

**Comment by drzraf** [comment]
#ansible-meeting, Thursday 15:00 UTC? I could be there some time.

**Comment by Shaps** [comment]
@drzraf Yes, 1500 UTC Thursday. 
Do you mind adding this as an item to the [Agenda](https://github.com/ansible/community/issues/450) ?

File: lib/ansible/modules/source_control/gitlab_service.py (hunk 22 above)
Comment (bcoca) [reply lib/ansible/modules/source_control/gitlab_service.py:348]: +10k to what @sivel states above

**Comment by Shaps** [comment]
@drzraf, the past few meetings were cancelled, do you think you can make it to today’s (Thursday 1500UTC) meeting? I should be around so can bring this up if you’re not around

**Comment by drzraf** [comment]
I got notified too late. Sorry.

File: lib/ansible/modules/source_control/gitlab_service.py (hunk 22 above)
Comment (drzraf) [reply lib/ansible/modules/source_control/gitlab_service.py:348]: Even if we were to autogenerate 30 modules, I doubt they would be merged either.
See my comment below :  Best next option is to simply drop the validation of this option so that no validation = no listing = less quality but rules respected. (and user will take care of their parameters themselves)

**Comment by Shaps** [comment]
@drzraf No worries, I had a chat with @sivel yesterday, and it has been decided in a previous meeting not to hold on the multiple services. That said it would be good if you could address the `dict()` which would make CI pass, I agree it would be misleading but as long as we have it written down in `DOCUMENTATION` we should be fine.

**Comment by drzraf** [comment]
Could be easier to wait for [this line](https://github.com/ansible/ansible/blob/devel/test/runner/lib/executor.py#L124) to be dropped rather messing around https://github.com/cwill747/alpine-python2.6 to run a Python interpreter released in 2008 and unsupported since October 2013...

**Comment by Akasurde** [comment]
Hi @drzraf, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>

**Comment by drzraf** [comment]
https://github.com/ansible-collections/community.general/pull/667
